sktree/tree/unsupervised/_unsup_criterion.pyx

```diff
@@ -398,3 +396,159 @@ cdef class TwoMeans(UnsupervisedCriterion):
         self.n_node_samples = end - start
         self.start = start
         self.end = end
+
+cdef class FastBIC(UnsupervisedCriterion):
+    """Placeholder for FastBIC docs
+    """
+
+    cdef double sum_of_squares(
+        self,
+        SIZE_t start,
+        SIZE_t end,
+        double mean,
+    ) nogil:
+        """Computes variance of feature vector from sample_indices[start:end].
+
+        See: https://en.wikipedia.org/wiki/Weighted_arithmetic_mean#Weighted_sample_variance.  # noqa
+
+        Parameters
+        ----------
+        start : SIZE_t
+            The start pointer
+        end : SIZE_t
+            The end pointer.
+        mean : double
+            The precomputed mean.
+
+        Returns
+        -------
+        ss : double
+            Sum of squares
+        """
+        cdef SIZE_t s_idx, p_idx        # initialize sample and pointer index
+        cdef double ss = 0.0            # sum-of-squares
+        cdef DOUBLE_t w = 1.0           # optional weight
+
+        # calculate variance for the sample_indices chosen start:end
+        for p_idx in range(start, end):
+            s_idx = self.sample_indices[p_idx]
+
+            # include optional weighted sum of squares
+            if self.sample_weight is not None:
+                w = self.sample_weight[s_idx]
+
+            ss += w * (self.Xf[s_idx] - mean) * (self.Xf[s_idx] - mean)
+        return ss
+
+    cdef double node_impurity(
+        self
+    ) nogil:
+        """Evaluate the impurity of the current node.
+
+        TODO: describe FastBIC
+
+        1. Prior = w = n/N
+        2. Mean = u
+        3. Vars = sig
+
+        impurity = -2*n*log(w) + n*log(2*pi*sig)
+        """
+        cdef double mean
+        cdef double impurity
+        cdef SIZE_t n_samples = self.n_samples #TODO: define n_samples
+        cdef SIZE_t n_node_samples = self.n_node_samples
+
+        # If calling without setting the
+        if self.Xf is None:
+            with gil:
+                raise MemoryError(
+                    'Xf has not been set yet, so one must call init_feature_vec.'
+                )
+
+        # first compute mean
+        mean = self.sum_total / n_node_samples
+
+        # then compute the variance of the cluster
+        sig = self.sum_of_squares(
+            self.start,
+            self.end,
+            mean
+        ) / self.weighted_n_node_samples
+
+        impurity = -2*n_node_samples*log(n_node_samles/n_samples) \ 
+                    + n_node_samples*log(2*cnp.pi*sig)
```

Comment:
Few comments regarding this line:

1. Does this compile? 
2. Also I find this a bit hard to read. Perhaps, you can simplify.
3. `n_node_samles/n_samples` isn't this just 1?
4. What equation are you following? This does not seem correct when I look at the paper.

Reply:
1. Definitely not especially with the typo here! I wanted to get your quick feedback on the theory side of things before I moved on
2. I'll work on the readability. Perhaps I could create a variable for the prior
3. Please correct me if I'm wrong, but in the paper, I think they are saying (n/N) where n is the current number of sample and N is the total number of samples. I still need to implement `n_samples` but that's kind of the idea that I was following
4. I'll post it in the comment below!

Reply:
`N` is the number of samples considered. At each node, only `n_node_samples` is considered. `n` in your case would be the number of samples determined by the split point. Since this is node_impurity, there is no split point, so `n=N`.

Reply:
This is super helpful, thank you so much